README
CARLA Simulator
===============

Thanks for downloading CARLA!

<http://carla.org/>

How to run CARLA
----------------

Launch a terminal in this folder and execute the simulator by running

```sh
 ./CarlaUE4.sh
```

this will launch a window with a view over the city. This is the "spectator"
view, you can fly around the city using the mouse and WASD keys, but you cannot
interact with the world in this view. The simulator is now running as a server,
waiting for a client app to connect and interact with the world.

Let's start by adding some live to the city, open a new terminal window and
execute

```sh
./generate_traffic.py -n 80
```

This adds 80 vehicles to the world driving in "autopilot" mode. Back to the
simulator window we should see these vehicles driving around the city. They will
keep driving randomly until we stop the script. Let's leave them there for now.

Now, it's nice and sunny in CARLA, but that's not a very interesting driving
condition. One of the cool features of CARLA is that you can control the weather
and lighting conditions of the world. We'll launch now a script that dynamically
controls the weather and time of the day, open yet another terminal window and
execute

```sh
./dynamic_weather.py
```

The city is now ready for us to drive, we can finally run

```sh
./manual_control.py
```

This should open a new window with a 3rd person view of a car, you can drive
this car with the WASD/arrow keys. Press 'h' to see all the options available.

For more details and running options please refer to our online documentation

<http://carla.readthedocs.io>


CHANGELOG
## CARLA 0.9.13

  * Added new **instance aware semantic segmentation** sensor `sensor.camera.instance_segmentation`
  * Added new API classes: `MaterialParameter`, `TextureColor` and  `TextureFloatColor` to encode texture data and field (normal map, diffuse, etc)
  * Added new API functions: `apply_color_texture_to_object`, `apply_float_color_texture_to_object` and `apply_textures_to_object` to paint objects in **runtime**
  * Added the option for users to set a **route** using RoadOption elements to a vehicle controlled by the Traffic Manager.
  * **Cache** now has an extra folder with current version of CARLA (so different cache per version)
  * Added **set_percentage_random_left_lanechange** and **set_percentage_random_right_lanechange**.
  * Improved handling of **collisions** in Traffic Manager when driving at **very high speeds**.
  * Added physical simulation to **vehicle doors**, capable of opening and closing
  * Added **open/close doors** feature for vehicles.
  * Added API functions to **3D vectors**: `squared_length`, `length`, `make_unit_vector`, `dot`, `dot_2d`, `distance`, `distance_2d`, `distance_squared`, `distance_squared_2d`, `get_vector_angle`
  * Added API functions to **2D vectors**: `squared_length`, `length`, `make_unit_vector`
  * Added a **seed** for better reproducibility of pedestrians
    - New API function `set_pedestrians_seed`
    - New parameter **--seedw** in generate_traffic.py script
  * Added missing dependency `libomp5` to **Release.Dockerfile**
  * Added API functions to interact with **pedestrian bones**:
    - `get_bones / set_bones`:  to get/set the bones of a pedestrian
    - `blend_pose`:  to blend a custom pose with current animation
    - `show_pose / hide_pose`:  to show or hide the custom pose
    - `get_pose_from_animation`:  to set the custom pose with the animation current frame
  * Added a new script in **PythonAPI/examples/draw_skeleton.py** to draw the bones of a pedestrian from client side
  * Improved **collision** detection of the Python agents
  * Added the new **VehicleLightStage** to the Traffic Manager to dynamically update the vehicle lights.
  * Added two new examples to **PythonAPI/util**: Conversion of OpenStreetMaps to OpenDRIVE maps `osm_to_xodr.py` and Extraction of map spawn points `extract_spawn_points.py`
  * Fixed the **import of props** without any map
  * Fixed **global route planner** crash when being used at maps without lane markings
  * Fixed bug causing the server to **sigsegv** when a vehicle collides an environment object in recording mode.
  * Fixed **RSSSensor**: made client side calculations threaded
  * Fixed **keep_right_rule** parameter.

## CARLA 0.9.12

  * Changed the resolution of the cached map in Traffic Manager from 0.1 to 5 meters
  * Fixed import sumo_integration module from other scripts
  * CARLA now is built with Visual Studio 2019 in Windows
  * Fixed bug causing the RoadOptions at the BehaviorAgent to not work as intended
  * Upgrading to Unreal Engine 4.26
  * Added generation attribute to vehicles and pedestrians
  * Added Lincoln 2020 vehicle dimensions for CarSim integration
  * Enabling the **no_delay** option to RPC and stream sockets
  * The special nomenclature to define roads (ROAD_ROAD), sidewalks (ROAD_SIDEWALK)... can be at any position of the asset name
  * Improved performance bencharmark script: sync, map and sensor selection, ...
  * Improved performance, destroyed PhysX state for vehicles when physics are disable
  * Improved parallelism of raycast sensors in system with large number of cores
  * Added 'visualize_multiple_sensors' example
  * Added 'check_lidar_bb' util script
  * Added optional flag to `client.replay_file()` `replay_sensors` to enable or disable the replaying the sensors
  * Improved manual_control: now cameras are set in relation with car size
  * Client required files are sent from the server to a local cache (OpenDRIVE, Traffic Manager...)
  * Added CHRONO library for vehicle dynamics simulation (https://projectchrono.org/)
    - Supported JSON vehicle definition
    - Unsupported collision dynamics
  * Added performance benchmarking section to documentation
  * Added API functions to traffic light actor `get_effect_waypoints()`, `get_light_boxes()` and `get_opendrive_id()`
  * Added API functions to world `get_traffic_lights_from_waypoint()`, `get_traffic_lights_in_junction` and `get_traffic_light_from_opendrive_id`
  * Added API parameters to `WorldSettings`: `tile_stream_distance` and `actor_active_distance`
  * Added API parameters and functions to `Osm2OdrSettings`: `proj_string`, `center_map`, `generate_traffic_lights`, `all_junctions_with_traffic_lights`, `set_osm_way_types`, and `set_traffic_light_excluded_way_types`
  * Added API function to enable car telemetry
  * CARLA is compatible with the last RoadRunner nomenclature for road assets
  * Fixed a bug when importing a FBX map with some **_** in the FBX name
  * Extended make import process for applying road painter materials (carla art tool)

Test_DoubleDQN.py
import random 
import sys
import glob
import os
from collections import deque
import numpy as np 
import cv2
import math
import time 
import tensorflow as tf
from keras.backend import set_session
from keras.models import load_model


try:
    sys.path.append(glob.glob('../carla/dist/carla-*%d.%d-%s.egg' % (
        sys.version_info.major,
        sys.version_info.minor,
        'win-amd64' if os.name == 'nt' else 'linux-x86_64'))[0])
except IndexError:
    pass
import carla

MODEL_PATH = 'models/Double_DQNmodel_Q2'
MEMORY_FRACTION = 0.8

IM_WIDTH = 640
IM_HEIGHT = 480
SHOW_PREVIEW  = False
SECONDS_PER_EPISODE = 30

class CarEnv:
    SHOW_CAM = SHOW_PREVIEW
    STEER_AMT = 1.0   ## full turn for every single time
    im_width = IM_WIDTH
    im_height = IM_HEIGHT
    front_camera = None
    action_space_size = len(ACTION)

    def __init__(self):
        self.client = carla.Client('localhost', 2000)
        self.client.set_timeout(200.0)
        # self.actor = carla.Actor
        
        #self.world = self.client.load_world('Town03')
        #self.map = self.world.get_map()   
        
        self.world = self.client.get_world()
        self.blueprint_library = self.world.get_blueprint_library()

        self.model_3 = self.blueprint_library.filter("model3")[0]  ## grab tesla model3 from library
        
        self.actions = [getattr(ACTIONS, action) for action in ACTION]

    def reset(self):
        self.collision_hist = []    
        self.actor_list = []
        self.laneInvasion_hist = []
        
        self.waypoints = self.client.get_world().get_map().generate_waypoints(distance=5.0)

        self.filtered_waypoints = []   
        i = 0
        for self.waypoint in self.waypoints:
            if(self.waypoint.road_id == 10):
                self.filtered_waypoints.append(self.waypoint)
                for i in range(len(self.filtered_waypoints)):
                    self.world.debug.draw_string(self.filtered_waypoints[i].transform.location, 'O', draw_shadow=False,
                                   color=carla.Color(r=0, g=255, b=0), life_time=40,
                                   persistent_lines=True)
                    i = i+1

        self.spawn_point = self.filtered_waypoints[1].transform

        self.spawn_point.location.z += 2
        self.vehicle = self.world.spawn_actor(self.model_3, self.spawn_point)  


        self.actor_list.append(self.vehicle)
        
        #Set Camera Attributes
        self.rgb_cam = self.blueprint_library.find('sensor.camera.rgb')
        self.rgb_cam.set_attribute("image_size_x", f"{self.im_width}")
        self.rgb_cam.set_attribute("image_size_y", f"{self.im_height}")
        self.rgb_cam.set_attribute("fov", f"110")  ## fov, field of view
        
        #Spawn and attach the camera
        transform = carla.Transform(carla.Location(x=2.5, z=0.7))
        self.sensor = self.world.spawn_actor(self.rgb_cam, transform, attach_to=self.vehicle)
        self.actor_list.append(self.sensor)
        self.sensor.listen(lambda data: self.process_img(data))
        self.vehicle.apply_control(carla.VehicleControl(throttle=0.0, brake=0.0))
        time.sleep(4)  # sleep to get things started and to not detect a collision when the car spawns/falls from sky.
        
        #Set up the collision sensor
        colsensor = self.world.get_blueprint_library().find('sensor.other.collision')
        self.colsensor = self.world.spawn_actor(colsensor, transform, attach_to=self.vehicle)
        self.actor_list.append(self.colsensor)
        self.colsensor.listen(lambda event: self.collision_data(event))
        
        #Set up the Lane Sensor
        lanesensor = self.world.get_blueprint_library().find('sensor.other.lane_invasion')
        self.lanesensor = self.world.spawn_actor(lanesensor, transform, attach_to=self.vehicle)
        self.actor_list.append(self.lanesensor)
        self.lanesensor.listen(lambda event: self.lane_invasion_data(event))

        while self.front_camera is None:  ## return the observation
            time.sleep(0.01)

        self.episode_start = time.time()

        self.vehicle.apply_control(carla.VehicleControl(brake=0.0, throttle=0.0))

        return self.front_camera

    def collision_data(self, event):
        self.collision_hist.append(event)
        
    def lane_invasion_data(self, event):
        #Lane Invasion Detected, return reward of -100
        self.laneInvasion_hist.append(event)
        

    def process_img(self, image):
        i = np.array(image.raw_data)
        #np.save("iout.npy", i)
        i2 = i.reshape((self.im_height, self.im_width, 4))
        i3 = i2[:, :, :3]
        if self.SHOW_CAM:
            cv2.imshow("",i3)
            cv2.waitKey(1)
        self.front_camera = i3  ## remember to scale this down between 0 and 1 for CNN input purpose


    def step(self, action):
        
        if action == 0:
            self.vehicle.apply_control(carla.VehicleControl(throttle=1.0, steer= 0.0 ))
            
        if action == 1:
            self.vehicle.apply_control(carla.VehicleControl(throttle=1.0, steer=1.0*self.STEER_AMT))
        if action == 2:
            self.vehicle.apply_control(carla.VehicleControl(throttle=1.0, steer=-1.0*self.STEER_AMT))
            
        v = self.vehicle.get_velocity()
        kmh = int(3.6 * math.sqrt(v.x**2 + v.y**2 + v.z**2))
      
        
        #Define rewards
        i = 2
        for i in range(2, len(self.filtered_waypoints)):


            if len(self.collision_hist) != 0:
                done = True
                reward = -300
            elif kmh < 30:
                done = False
                reward = -5
            elif carla.Location.distance(carla.Actor.get_location(self.actor_list[0]), self.filtered_waypoints[i].transform.location) == 0:
                done = False
                reward = 25
            elif len(self.laneInvasion_hist) != 0:
                done = False
                reward = -10
            else:
                done = False
                reward = 30
            i = i + 1

            if self.episode_start + SECONDS_PER_EPISODE < time.time():  ## when to stop
                done = True

            return self.front_camera, reward, done, None
            
if __name__ == '__main__':

    #Set GPU Options 
    gpu_options = tf.compat.v1.GPUOptions(per_process_gpu_memory_fraction=MEMORY_FRACTION)
    tf.compat.v1.keras.backend.set_session(tf.compat.v1.Session(config=tf.compat.v1.ConfigProto(gpu_options=gpu_options)))
    
    #Load the Model
    model = load_model(MODEL_PATH)
    print("Loading Model from: ", MODEL_PATH)
    
    #Create the environment for the car 
    env = CarEnv()
    
    #For agent speed measurements, keep last 60 frames
    fps_counter = deque(maxlen=60)
    
    #Initialize the predictions 
    model.predict(np.ones((1, IM_HEIGHT, IM_WIDTH, 3)))
    
    #Main loop 
    while True:

        print('Restarting Episodes')

        current_state = env.reset()
        env.collision_hist = []

        done = False

        while True:
            step_start = time.time()
        
            qs = model.predict(np.array(current_state).reshape(-1, *current_state.shape)/255)[0]
            action = np.argmax(qs)
        
            new_state, reward, done, _ = env.step(action)
        
            current_state = new_state
        
            if done:
                break
        
            frame_time = time.time() - step_start
            fps_counter.append(frame_time)
            print(f'Agent: {len(fps_counter)/sum(fps_counter):>4.1f} FPS | Action: {action}')
        
        #Destroy an actor at the end of an episode
        for actor in env.actor_list:
            actor.destroy()

Improved_DQN.py
import glob
import os
import sys
import random
import time
import sys
import numpy as np
import cv2
import math
import matplotlib.pyplot as plt
from collections import deque
import tensorflow as tf
from keras.applications.xception import Xception
#from keras.applications.MobileNetV2 import MobileNetV2
from keras.models import Sequential, Model
from keras.layers import Dense, GlobalAveragePooling2D, Input, Concatenate, Conv2D, AveragePooling2D, Activation, Flatten
from keras.optimizers import Adam
from keras.models import Model
from keras.callbacks import TensorBoard
import tensorflow.keras.backend as backend
from tensorflow.keras.callbacks import ModelCheckpoint
from threading import Thread
from tqdm import tqdm
from keras.models import load_model


#Header for any CARLA script, this is for the CARLA python library 
try:
    sys.path.append(glob.glob('../carla/dist/carla-*%d.%d-%s.egg' % (
        sys.version_info.major,
        sys.version_info.minor,
        'win-amd64' if os.name == 'nt' else 'linux-x86_64'))[0])
except IndexError:
    pass
import carla

#Set parameters for the simulation 
IM_WIDTH = 960
IM_HEIGHT = 480
SECONDS_PER_EPISODE = 25
REPLAY_MEMORY_SIZE = 5_000

MIN_REPLAY_MEMORY_SIZE = 1_000
MINIBATCH_SIZE = 16
PREDICTION_BATCH_SIZE = 1
TRAINING_BATCH_SIZE = MINIBATCH_SIZE // 4
UPDATE_TARGET_EVERY = 5  #used to be 10
MODEL_NAME = "Improved_DQN"

MEMORY_FRACTION = 0.8
MIN_REWARD = 100



EPISODES = 8000

DISCOUNT = 0.99
epsilon = 1
EPSILON_DECAY = 0.95 ## 0.9975 99975
MIN_EPSILON = 0.001

AGGREGATE_STATS_EVERY = 5  ## checking per 5 episodes
SHOW_PREVIEW  = False    ## for debugging purpose

#Define the model path 
MODEL_PATH = 'models/Improved_DQNmodel'

#Class for the extended actions
class ACTIONS:
    forward = 0
    left = 1
    right = 2
    forward_left = 3
    forward_right = 4
    brake = 5
    brake_left = 6
    brake_right = 7
    

ACTION_CONTROL = {
    0: [1, 0, 0],
    1: [0, 0, -1],
    2: [0, 0, 1],
    3: [1, 0, -1],
    4: [1, 0, 1],
    5: [0, 1, 0],
    6: [0, 1, -1],
    7: [0, 1, 1]
}

ACTIONS_NAMES = {
    0: 'forward',
    1: 'left',
    2: 'right',
    3: 'forward_left',
    4: 'forward_right',
    5: 'brake',
    6: 'brake_left',
    7: 'brake_right'
}

ACTION = ['forward', 'left', 'right', 'forward_left', 'forward_right', 'brake', 'brake_left', 'brake_right']

#Class to create the Car Environment
class CarEnv:
    SHOW_CAM = SHOW_PREVIEW
    STEER_AMT = 1.0   ## full turn for every single time
    im_width = IM_WIDTH
    im_height = IM_HEIGHT
    front_camera = None
    action_space_size = len(ACTION)

    def __init__(self):
        self.client = carla.Client('localhost', 2000) #This is what listens to the server
        self.client.set_timeout(200.0)
        # self.actor = carla.Actor
        #self.world = self.client.load_world('Town03')
        #self.map = self.world.get_map()   ## added for map creating
        self.world = self.client.get_world() #Get the blueprint for the world, i.e. the map
        self.blueprint_library = self.world.get_blueprint_library()

        self.model_3 = self.blueprint_library.filter("model3")[0]  #tesla model3 from library
        
        self.actions = [getattr(ACTIONS, action) for action in ACTION] #Get the number of actions from the action class

    def reset(self):
        self.collision_hist = []    
        self.actor_list = []
        self.laneInvasion_hist = []
        
        self.waypoints = self.client.get_world().get_map().generate_waypoints(distance=5.0)

        self.filtered_waypoints = []   
        i = 0
        for self.waypoint in self.waypoints:
            if(self.waypoint.road_id == 10):
                self.filtered_waypoints.append(self.waypoint)
                for i in range(len(self.filtered_waypoints)):
                    self.world.debug.draw_string(self.filtered_waypoints[i].transform.location, 'O', draw_shadow=False,
                                   color=carla.Color(r=0, g=255, b=0), life_time=40,
                                   persistent_lines=True)
                    i = i+1

        self.spawn_point = self.filtered_waypoints[1].transform

        self.spawn_point.location.z += 2
        self.vehicle = self.world.spawn_actor(self.model_3, self.spawn_point)  ## changed for adding waypoints


        self.actor_list.append(self.vehicle)
        
        #Set RGB Camera Attributes
        self.rgb_cam = self.blueprint_library.find('sensor.camera.rgb')
        self.rgb_cam.set_attribute("image_size_x", f"{self.im_width}")
        self.rgb_cam.set_attribute("image_size_y", f"{self.im_height}")
        self.rgb_cam.set_attribute("fov", f"110")  ## fov, field of view
        
        #Spawn and attach the camera
        transform = carla.Transform(carla.Location(x=2.5, z=0.7))
        self.sensor = self.world.spawn_actor(self.rgb_cam, transform, attach_to=self.vehicle)
        self.actor_list.append(self.sensor)
        self.sensor.listen(lambda data: self.process_img(data))
        self.vehicle.apply_control(carla.VehicleControl(throttle=0.0, brake=0.0)) 
        time.sleep(4)  # sleep to get things started and to not detect a collision when the car spawns/falls from sky.
        
        #Set up the collision sensor
        colsensor = self.world.get_blueprint_library().find('sensor.other.collision')
        self.colsensor = self.world.spawn_actor(colsensor, transform, attach_to=self.vehicle)
        self.actor_list.append(self.colsensor)
        self.colsensor.listen(lambda event: self.collision_data(event))
        
        #Set up the Lane Sensor
        lanesensor = self.world.get_blueprint_library().find('sensor.other.lane_invasion')
        self.lanesensor = self.world.spawn_actor(lanesensor, transform, attach_to=self.vehicle)
        self.actor_list.append(self.lanesensor)
        self.lanesensor.listen(lambda event: self.lane_invasion_data(event))

        while self.front_camera is None:  ## return the observation
            time.sleep(0.01)

        self.episode_start = time.time()

        self.vehicle.apply_control(carla.VehicleControl(brake=0.0, throttle=0.0))

        return self.front_camera

    def collision_data(self, event):
        self.collision_hist.append(event) #Look to see if a collision occured 
        
    def lane_invasion_data(self, event):
        #Look to see if a lane invasion occured 
        self.laneInvasion_hist.append(event)
        

    def process_img(self, image):
        #Function to process the image to be passed to the CNN
        i = np.array(image.raw_data)
        #np.save("iout.npy", i)
        i2 = i.reshape((self.im_height, self.im_width, 4))
        i3 = i2[:, :, :3]
        if self.SHOW_CAM:
            cv2.imshow("",i3)
            cv2.waitKey(1)
        self.front_camera = i3  ## remember to scale this down between 0 and 1 for CNN input purpose


    def step(self, action):
        #Function to apply control to the vehicle and get the reward, i.e. this is the function that takes the action 
        
        #Apply the action, an action must be taken every step 
        self.vehicle.apply_control(carla.VehicleControl(throttle=ACTION_CONTROL[self.actions[action]][0], steer=ACTION_CONTROL[self.actions[action]][2]*self.STEER_AMT, brake=ACTION_CONTROL[self.actions[action]][1]))
           
        v = self.vehicle.get_velocity()
        kmh = int(3.6 * math.sqrt(v.x**2 + v.y**2 + v.z**2))
      
        
        #Define rewards
        i = 2
        for i in range(2, len(self.filtered_waypoints)):


            if len(self.collision_hist) != 0:
                done = True
                reward = -300
            elif kmh < 30:
                done = False
                reward = -15
            elif carla.Location.distance(carla.Actor.get_location(self.actor_list[0]), self.filtered_waypoints[i].transform.location) == 0:
                done = False
                reward = 25
            elif len(self.laneInvasion_hist) != 0:
                done = False
                reward = -10
            else:
                done = False
                reward = 30
            i = i + 1

            if self.episode_start + SECONDS_PER_EPISODE < time.time():  ## when to stop
                done = True

            return self.front_camera, reward, done, None


#Class for the DQN Agent
class DQNAgent:
    def __init__(self):


        ## replay_memory is used to remember the sized previous actions, and then fit our model of this amout of memory by doing random sampling
        self.replay_memory = deque(maxlen=REPLAY_MEMORY_SIZE)   ## batch step
        self.tensorboard = ModifiedTensorBoard(log_dir=f"logs/{MODEL_NAME}-{int(time.time())}")

        self.target_update_counter = 0  # will track when it's time to update the target model
       
        #self.model = self.create_model()
        ## target model (this is what we predict against every step)
        self.target_model = self.create_model()
        
        #If partial training done before, load the model instead of creating it
        self.model = load_model(MODEL_PATH)
        
        self.target_model.set_weights(self.model.get_weights())
        
        self.model.summary()

        self.terminate = False  # Should we quit?
        self.last_logged_episode = 0
        self.training_initialized = False  # waiting for TF to get rolling

    #function to create model for the CNN
    def create_model(self):
        ## input: RGB data, should be normalized when coming into CNN

        base_model = tf.keras.applications.MobileNetV2(weights='imagenet', include_top=False, input_shape=(IM_HEIGHT, IM_WIDTH,3)) 
        x = base_model.output
        x = AveragePooling2D(pool_size=(15,30))(x)
        x = Conv2D(filters = 1000, kernel_size=(1,1), activation = 'relu')(x)
        x = Flatten()(x) 

        predictions = Dense(8, activation="linear")(x)  ## output layer include 8 nuros, representing 8 actions
        model = Model(inputs=base_model.input, outputs=predictions)
        model.compile(loss="mse", optimizer=Adam(learning_rate=0.0001), metrics=["accuracy"])          
        
        return model

    ## function handler
    # Adds step's data to a memory replay array
    # (observation space, action, reward, new observation space, done)= (current_state, action, reward, new_state, done)
    def update_replay_memory(self, transition):
        self.replay_memory.append(transition)

    #Function for training
    def train(self):

        ## starting training only if certain number of samples is already saved
        if len(self.replay_memory) < MIN_REPLAY_MEMORY_SIZE:
            return

        ## if we do have the proper amount of data to train, we need to randomly select the data we want to train off from our memory
        minibatch = random.sample(self.replay_memory, MINIBATCH_SIZE)

        ## get current states from minibatch and then get Q values from NN model
        ## transition is being defined by this: transition = (current_state, action, reward, new_state, done)
        current_states = np.array([transition[0] for transition in minibatch])/255


        ## This is the crazyly changed model:
       
        current_qs_list = self.model.predict(current_states, PREDICTION_BATCH_SIZE)    ## changed
        

        ## This is normal model
        new_current_states = np.array([transition[3] for transition in minibatch])/255
        future_qs_list = self.target_model.predict(new_current_states, PREDICTION_BATCH_SIZE)

        ## image data(normalized RGB data): input
        X = []
        ## action we take(Q values): output
        y = []

        ## calculate Q values for the next step based on Qnew equation
        ## index = step
        for index, (current_state, action, reward, new_state, done) in enumerate(minibatch):
            if not done:
                max_future_q = np.max(future_qs_list[index])
                new_q = reward + DISCOUNT * max_future_q
            else:
                new_q = reward

            current_qs = current_qs_list[index]
            current_qs[action] = new_q     ## Q for the action that we took is now equal to the new Q value

            X.append(current_state)  ## image we have 
            y.append(current_qs)  ## Q value we have

        ## only trying to log per episode, not actual training step, so we're going to use the below to keep track
        log_this_step = False
        if self.tensorboard.step > self.last_logged_episode:
            log_this_step = True
            self.last_log_episode = self.tensorboard.step

        #Filepath
        filepath = 'models/Improved_DQNmodel'
        #Checkpoint
        checkpoint = ModelCheckpoint(filepath, monitor='val_accuracy', verbose = 1, save_best_only = True, mode = 'max')
        callbacks_list = [checkpoint]
        
        ## fit our model
        ## setting the tensorboard callback, only if log_this_step is true. If it's false, then we'll still fit, we just wont log to TensorBoard.
        self.model.fit(np.array(X)/255, np.array(y), batch_size=TRAINING_BATCH_SIZE, verbose=0, shuffle=False, callbacks=[self.tensorboard])

        ## updating to determine if we want to update target_model 
        if log_this_step:
            self.target_update_counter += 1

        # If counter reaches set value, update target network with weights of main network
        if self.target_update_counter > UPDATE_TARGET_EVERY:
            self.target_model.set_weights(self.model.get_weights())
            self.target_update_counter = 0

    def get_qs(self, state):
        q_out = self.model.predict(np.array(state).reshape(-1, *state.shape)/255)[0]
        return q_out

        ## first to train to some nonsense. just need to get a quicl fitment because the first training and predication is slow
    def train_in_loop(self):
        X = np.random.uniform(size=(1, IM_HEIGHT, IM_WIDTH, 3)).astype(np.float32)
        y = np.random.uniform(size=(1, 8)).astype(np.float32)
        self.model.fit(X,y, verbose=False, batch_size=1)

        self.training_initialized = True

        while True:
            if self.terminate:
                return
            self.train()
            time.sleep(0.01)
        

#Class that saves the results using a library called TensorBoard
class ModifiedTensorBoard(TensorBoard):

    # Overriding init to set initial step and writer (we want one log file for all .fit() calls)
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.step = 1
        # self.writer = tf.summary.FileWriter(self.log_dir)
        self.writer = tf.summary.create_file_writer('ImprovedDQN-log')
        

    # Overriding this method to stop creating default log writer
    def set_model(self, model):
        pass

    # Overrided, saves logs with our step number
    # (otherwise every .fit() will start writing from 0th step)
    def on_epoch_end(self, epoch, logs=None):
        self.update_stats(**logs)

    # Overrided
    # We train for one batch only, no need to save anything at epoch end
    def on_batch_end(self, batch, logs=None):
        pass

    # Overrided, so won't close writer
    def on_train_end(self, _):
        pass

    def _write_logs(self, logs, index):
        with self.writer.as_default():
            for name, value in logs.items():
                tf.summary.scalar(name, value, step=index)
                self.step += 1
                self.writer.flush()

    # Custom method for saving own metrics
    # Creates writer, writes custom metrics and closes writer
    def update_stats(self, **stats):
        self._write_logs(stats, self.step)
      

#Main function 
if __name__ == '__main__':
    FPS = 60
    # For stats
    ep_rewards = [-200]

    # For more repetitive results
    random.seed(1)
    np.random.seed(1)
    tf.random.set_seed(1)

    # Memory fraction, used mostly when training multiple agents
    gpu_options = tf.compat.v1.GPUOptions(per_process_gpu_memory_fraction=MEMORY_FRACTION)
    tf.compat.v1.keras.backend.set_session(tf.compat.v1.Session(config=tf.compat.v1.ConfigProto(gpu_options=gpu_options)))

    # Create models folder, this is where the model will go 
    if not os.path.isdir('models'):
        os.makedirs('models')

    # Create agent and environment
    agent = DQNAgent()
    env = CarEnv()


    # Start training thread and wait for training to be initialized
    trainer_thread = Thread(target=agent.train_in_loop, daemon=True)
    trainer_thread.start()
    while not agent.training_initialized:
        time.sleep(0.01)

    #Get the Q value from the Q function 
    agent.get_qs(np.ones((env.im_height, env.im_width, 3)))
    
    
    # Iterate over episodes
    for episode in tqdm(range(1, EPISODES + 1), unit='episodes'):
        try:

            env.collision_hist = []

            # Update tensorboard step every episode
            agent.tensorboard.step = episode

            # Restarting episode - reset episode reward and step number
            episode_reward = 0
            score = 0
            step = 1

            # Reset environment and get initial state
            current_state = env.reset()

            # Reset flag and start iterating until episode ends
            done = False
            episode_start = time.time()
            # Play for given number of seconds only
            while True:

                # np.random.random() will give us the random number between 0 and 1. If this number is greater than our randomness variable,
                # we will get Q values baed on tranning, but otherwise, we will go random actions.
                if np.random.random() > epsilon:
                    # Get action from Q table
                    action = np.argmax(agent.get_qs(current_state))
                else:
                    # Get random action
                    action = np.random.randint(0, 8)
                    # This takes no time, so we add a delay matching 60 FPS (prediction above takes longer)
                    time.sleep(1/FPS)

                new_state, reward, done, _ = env.step(action)

                # Transform new continous state to new discrete state and count reward
                episode_reward += reward

                # Every step we update replay memory
                agent.update_replay_memory((current_state, action, reward, new_state, done))

                current_state = new_state
                step += 1

                if done:
                    break


            # End of episode - destroy agents
            for actor in env.actor_list:
                actor.destroy()
            
            scores.append(score)
            avg_scores.append(np.mean(scores[-10:]))
            
            # Append episode reward to a list and log stats (every given number of episodes)
            ep_rewards.append(episode_reward)
            if not episode % AGGREGATE_STATS_EVERY or episode == 1:        ## every show_stats_every, which is 10 right now, show and save teh following
                average_reward = sum(ep_rewards[-AGGREGATE_STATS_EVERY:])/len(ep_rewards[-AGGREGATE_STATS_EVERY:])
                min_reward = min(ep_rewards[-AGGREGATE_STATS_EVERY:])
                max_reward = max(ep_rewards[-AGGREGATE_STATS_EVERY:])
                agent.tensorboard.update_stats(reward_avg=average_reward, reward_min=min_reward, reward_max=max_reward, epsilon=epsilon)

                # Save model, but only when min reward is greater or equal a set value
                if average_reward >= MIN_REWARD:
                    agent.model.save('Improved_DQNmodel')

            # Decay epsilon
            if epsilon > MIN_EPSILON:
                epsilon *= EPSILON_DECAY
                epsilon = max(MIN_EPSILON, epsilon)
                
        except KeyboardInterrupt:
            print('KeyboardInterrupt: Saving agent model')
            agent.terminate = True
            trainer_thread.join()
            agent.model.save('models/Improved_DQNmodel')

    # Set termination flag for training thread and wait for it to finish
    agent.terminate = True
    trainer_thread.join()
    agent.model.save('models/Improved_DQNmodel')
    
    